plugins/rest/rest_auth.go

```diff
@@ -113,56 +121,171 @@ type tokenEndpointResponse struct {
 // oauth2ClientCredentialsAuthPlugin represents authentication via a bearer token in the HTTP Authorization header
 // obtained through the OAuth2 client credentials flow
 type oauth2ClientCredentialsAuthPlugin struct {
-	TokenURL     string    `json:"token_url"`
-	ClientID     string    `json:"client_id"`
-	ClientSecret string    `json:"client_secret"`
-	Scopes       *[]string `json:"scopes,omitempty"`
-
-	tokenCache    *oauth2Token
-	tlsSkipVerify bool
+	GrantType    string                 `json:"grant_type"`
+	TokenURL     string                 `json:"token_url"`
+	ClientID     string                 `json:"client_id"`
+	ClientSecret string                 `json:"client_secret"`
+	SigningKeyID string                 `json:"signing_key"`
+	Claims       map[string]interface{} `json:"additional_claims"`
+	IncludeJti   bool                   `json:"include_jti_claim"`
+	Scopes       []string               `json:"scopes,omitempty"`
+
+	signingKey       *keys.Config
+	signingKeyParsed interface{}
+	tokenCache       *oauth2Token
+	tlsSkipVerify    bool
 }
 
 type oauth2Token struct {
 	Token     string
 	ExpiresAt time.Time
 }
 
+func (ap *oauth2ClientCredentialsAuthPlugin) createAuthJWT(claims map[string]interface{}, signingKey interface{}) (*string, error) {
+	now := time.Now()
+	baseClaims := map[string]interface{}{
+		"iat": now.Unix(),
+		"exp": now.Add(10 * time.Minute).Unix(),
+	}
+	if claims == nil {
+		claims = make(map[string]interface{})
+	}
+	for k, v := range baseClaims {
+		claims[k] = v
+	}
+
+	if len(ap.Scopes) > 0 {
+		claims["scope"] = strings.Join(ap.Scopes, " ")
+	}
+
+	if ap.IncludeJti {
+		jti, err := uuid.New(rand.Reader)
+		if err != nil {
+			return nil, err
+		}
+		claims["jti"] = jti
+	}
+
+	payload, err := json.Marshal(claims)
+	if err != nil {
+		return nil, err
+	}
+
+	jwsHeaders := []byte(fmt.Sprintf(`{"typ":"JWT","alg":"%v"}`, ap.signingKey.Algorithm))
+	jwsCompact, err := jws.SignLiteral(payload, jwa.SignatureAlgorithm(ap.signingKey.Algorithm), signingKey, jwsHeaders)
+	if err != nil {
+		return nil, err
+	}
+	jwt := string(jwsCompact[:])
```

Comment:
❓ What does the `[:]` give us?

Reply:
Honestly not sure - it was a pattern inspired/stolen from the JWS lib without much further thought, like here https://github.com/open-policy-agent/opa/blob/master/internal/jwx/jws/jws.go#L157

Reply:
Convert an array (fixed size) to a slice (dynamic sized)

Reply:
💡 Cool, I wasn't aware of the array-to-slice functionality of this. I was confused because it seems like it's returning a byte slice already,

https://github.com/open-policy-agent/opa/blob/f40e9350a9f3cd2d4bc5b0553a912a743ccdba99/internal/jwx/jws/jws.go#L40

So, `string(jwsCompact[:])` would do the same thing as `string(jwsCompact)`, wouldn't it? 🤔

Reply:
@srenatus Correct, we don't need [:]. It adds no value at all.

Reply:
OK - removed here and in tokens.go .. still curious about the original intent for this.